### PR TITLE
Introduce feature gates for the scheduler component

### LINKED ISSUE #701: Add featuregate support
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by iawia002**
Signed-off-by: iawia002 <z2d@jifangcheng.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Introduce the feature gates package to allow users to control whether to enable/disable some global features, this can also be used in other components in the future.

**Which issue(s) this PR fixes**:
Fixes #701

**Special notes for your reviewer**:

I directly removed the `--failover` flag. I don’t know if we need to ensure compatibility in the current version.

**Does this PR introduce a user-facing change?**:

```release-note
Introduce feature gates for the scheduler component.
```

/cc @RainbowMango

### CODE REVIEW

**Comment by RainbowMango** [comment]
That's interesting. I'll look at it later. Thanks!
/assign

File: pkg/features/features.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package features
```
[the remark this answers is not in the document]
Comment (iawia002) [reply pkg/features/features.go]: Any update on this?

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/features/features.go]: Oh...sorry, I didn't have enough time to investigate it. I guess by this week should be ok.

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/features/features.go]: > I did some basic research, the featuregate package does meet our needs, but there is a small problem, it has two features by default and cannot be configured, AllAlpha and AllBeta don't really work, can we accept this?

Yes, the package is not suitable for reuse by our project. 
So, how about implementing a new package? we can borrow some ideas/methods from k/k.

[the remark this answers is not in the document]
Comment (iawia002) [reply pkg/features/features.go]: > So, how about implementing a new package? we can borrow some ideas/methods from k/k.

The previous version satisfies our needs. I made some changes based on the tidb-operator features package which is also based on the k/k code. Maybe I can add some comments to explain the original source of this code?

[the remark this answers is not in the document]
Comment (RainbowMango) [reply pkg/features/features.go]: I tend to based on k/k. And add a note to the package, like
```
This code is lifted from Kubernetes codebase. 
For reference: https://github.com/kubernetes/kubernetes/blob/v1.22.2/staging/src/k8s.io/component-base/featuregate/feature_gate.go
```

[the remark this answers is not in the document]
Comment (iawia002) [reply pkg/features/features.go]: Sure, will update it soon.

[the remark this answers is not in the document]
Comment (iawia002) [reply pkg/features/features.go]: > `AllAlpha` and `AllBeta` don't really work, can we accept this?

Oh I missed something when reading the code, `AllAlpha` and `AllBeta` do work, it is a global toggle for alpha/beta features, do we need those two features? If we do, we don't need to copy the code then.

File: pkg/features/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package scheduler
```
Comment (RainbowMango) [inline pkg/features/scheduler/scheduler.go:1]: I suggest renaming the file to `pkg/features/features.go`.

Comment (iawia002) [reply pkg/features/scheduler/scheduler.go]: Sounds reasonable, I have updated.

**Pull request merged**